firstpartydata/first_party_data.go

```diff
@@ -0,0 +1,376 @@
+package firstpartydata
+
+import (
+	"fmt"
+	"github.com/evanphx/json-patch"
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/util/jsonutil"
+)
+
+const (
+	siteKey = "site"
+	appKey  = "app"
+	userKey = "user"
+	dataKey = "data"
+
+	userDataKey        = "userData"
+	appContentDataKey  = "appContentData"
+	siteContentDataKey = "siteContentData"
+)
+
+func ExtractGlobalFPDData(req *openrtb_ext.RequestWrapper) (map[string][]byte, error) {
+	//If {site,app,user}.ext.data exists, collect it and remove {site,app,user}.ext.data
+
+	fpdReqData := make(map[string][]byte, 3)
+
+	siteExt, err := req.GetSiteExt()
+	if err != nil {
+		return nil, err
+	}
+
+	if siteExt != nil && len(siteExt.GetExt()[dataKey]) > 0 {
+		fpdReqData[siteKey] = siteExt.GetExt()[dataKey]
+		newSiteExt, err := jsonutil.DropElement(req.Site.Ext, dataKey)
+		if err != nil {
+			return nil, err
+		}
+		req.Site.Ext = newSiteExt
+	}
+
+	appExt, err := req.GetAppExt()
+	if err != nil {
+		return nil, err
+	}
+	if appExt != nil && len(appExt.GetExt()[dataKey]) > 0 {
+		fpdReqData[appKey] = appExt.GetExt()[dataKey]
+		newAppExt, err := jsonutil.DropElement(req.App.Ext, dataKey)
+		if err != nil {
+			return nil, err
+		}
+		req.App.Ext = newAppExt
+	}
+
+	userExt, err := req.GetUserExt()
+	if err != nil {
+		return nil, err
+	}
+	if req.User != nil && len(userExt.GetExt()[dataKey]) > 0 {
+		fpdReqData[userKey] = userExt.GetExt()[dataKey]
+		newUserExt, err := jsonutil.DropElement(req.User.Ext, dataKey)
+		if err != nil {
+			return nil, err
+		}
+		req.User.Ext = newUserExt
+	}
+
+	return fpdReqData, nil
+}
+
+func ExtractOpenRtbGlobalFPD(bidRequest *openrtb2.BidRequest) map[string][]openrtb2.Data {
+	//Extract and delete user.data and {app/site}.content.data from request
+
+	openRtbGlobalFPD := make(map[string][]openrtb2.Data, 3)
+	if bidRequest.User != nil && len(bidRequest.User.Data) > 0 {
+		openRtbGlobalFPD[userDataKey] = bidRequest.User.Data
+		bidRequest.User.Data = nil
+	}
+
+	if bidRequest.Site != nil && bidRequest.Site.Content != nil && len(bidRequest.Site.Content.Data) > 0 {
+		openRtbGlobalFPD[siteContentDataKey] = bidRequest.Site.Content.Data
+		bidRequest.Site.Content.Data = nil
+	}
+
+	if bidRequest.App != nil && bidRequest.App.Content != nil && len(bidRequest.App.Content.Data) > 0 {
+		openRtbGlobalFPD[appContentDataKey] = bidRequest.App.Content.Data
+		bidRequest.App.Content.Data = nil
+	}
+
+	return openRtbGlobalFPD
+
+}
+
+func ResolveFPDData(bidRequest *openrtb2.BidRequest, fpdBidderConfigData map[openrtb_ext.BidderName]*openrtb_ext.ORTB2, globalFPD map[string][]byte, openRtbGlobalFPD map[string][]openrtb2.Data, biddersWithGlobalFPD []string) (map[openrtb_ext.BidderName]*openrtb_ext.ORTB2, []error) {
+	errL := []error{}
```

Comment:
Please declare with `var errL []error` to avoid an allocation if there are no errors.

Reply:
Refactored